import org.jetbrains.kotlin.formver.plugin.Pure
import org.jetbrains.kotlin.formver.plugin.DumpExpEmbeddings

fun <!VIPER_TEXT!>emptyFunction<!>() { }

@Pure
fun <!VIPER_TEXT!>emptyAnnotatedFunction<!>() { }
